### PR TITLE
feat: Support determining extensions from names like `foo.parquet.snappy` as well as `foo.parquet`

### LINKED ISSUE #7954: When the file extension is changed, read silently fails.
### Describe the bug

When reading from a parquet file, if the file extension doesn't happen to match the default configuration, an empty `DataFrame` is produced. I believe this is easily encountered as a beginner (reading a Parquet file using the default read options is one of the first things I tried), and the lack of feedback makes this a challenge to debug.

### To Reproduce

```
//! ```cargo
//! [dependencies]
//! anyhow = "1.0"
//! datafusion = "32.0"
//! tokio = { version = "1.33", features = ["macros", "rt-multi-thread"] }
//! ```
use std::sync::Arc;

use datafusion::arrow::array::{Float32Array, Int32Array};
use datafusion::arrow::datatypes::{DataType, Field, Schema};
use datafusion::arrow::record_batch::RecordBatch;
use datafusion::dataframe::DataFrameWriteOptions;
use datafusion::parquet::basic::Compression;
use datafusion::parquet::file::properties::WriterProperties;
use datafusion::prelude::*;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let ctx = SessionContext::new();

    // Make up a new dataframe.
    let write_df = ctx.read_batch(RecordBatch::try_new(
        Arc::new(Schema::new(vec![
            Field::new("purchase_id", DataType::Int32, false),
            Field::new("price", DataType::Float32, false),
            Field::new("quantity", DataType::Int32, false),
        ])),
        vec![
            Arc::new(Int32Array::from(vec![1, 2, 3, 4, 5])),
            Arc::new(Float32Array::from(vec![1.12, 3.40, 2.33, 9.10, 6.66])),
            Arc::new(Int32Array::from(vec![1, 3, 2, 4, 3])),
        ],
    )?)?;

    write_df
        .write_parquet(
            "output.parquet.snappy",
            DataFrameWriteOptions::new().with_single_file_output(true),
            Some(
                WriterProperties::builder()
                    .set_compression(Compression::SNAPPY)
                    .build(),
            ),
        )
        .await?;

    let read_df = ctx
        .read_parquet(
            "output.parquet.snappy",
            ParquetReadOptions {
                // If this line is uncommented, the read will be successful.
                // file_extension: "parquet.snappy",
                ..Default::default()
            },
        )
        .await?;

    read_df.show().await?;

    Ok(())
}
```

Gives output:

```
$ cargo run
   Compiling bug_repro_open_no_error v0.1.0 (/home/jacob/src/gro-sandbox/user/jsimpson/app/bug_repro_open_no_error)
    Finished dev [unoptimized + debuginfo] target(s) in 7.56s
     Running `target/debug/bug_repro_open_no_error`
++
++
```

### Expected behavior

I would like to get an error if the read fails, rather than an empty `DataFrame`.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Closes #7954

### CODE REVIEW

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -849,6 +849,33 @@ impl SessionContext {
         let table_paths = table_paths.to_urls()?;
         let session_config = self.copied_config();
         let listing_options = options.to_listing_options(&session_config);
+
+        let option_extention = listing_options.file_extension.clone();
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:853]: please rename is to `extension` not extention

File: datafusion/core/src/execution/context/mod.rs [hunk 2]
```diff
@@ -849,6 +849,33 @@ impl SessionContext {
...
+        let filename = table_paths[0].prefix().filename();
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:854]: we prob should check `table_paths` its not empty before accessing the element by index

File: datafusion/core/src/execution/context/mod.rs [hunk 3]
```diff
@@ -849,6 +849,33 @@ impl SessionContext {
...
+            "".to_owned()
+        };
+        // some the file extension might be started with "." and some not
+        let extention_alternative = ".".to_string() + extention.as_str();
+
+        if option_extention != extention
+            && option_extention != extention_alternative
+            && !extention.is_empty()
+        {
+            return Err(DataFusionError::Execution(format!(
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:873]: please replace with `exec_err!` macros

File: datafusion/core/src/execution/context/mod.rs [hunk 4]
```diff
@@ -849,6 +849,33 @@ impl SessionContext {
         let table_paths = table_paths.to_urls()?;
         let session_config = self.copied_config();
         let listing_options = options.to_listing_options(&session_config);
+
+        let option_extention = listing_options.file_extension.clone();
+        let filename = table_paths[0].prefix().filename();
+        let extention = if let Some(filename) = filename {
+            let parts: Vec<&str> = filename.split('.').collect();
+
+            if parts.len() > 1 {
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:858]: its likely can be simplified with `.and_then` or `.map`

File: datafusion/core/src/execution/context/parquet.rs [hunk 5]
```diff
@@ -132,6 +137,49 @@ mod tests {
...
+
+        let read_df = ctx
+            .read_parquet(
+                "output.parquet.snappy",
+                ParquetReadOptions {
+                    ..Default::default()
+                },
+            )
+            .await;
+        assert!(read_df.is_err());
```
Comment (comphead) [inline datafusion/core/src/execution/context/parquet.rs:178]: Please test the error text message as well.

File: datafusion/core/src/execution/context/parquet.rs [hunk 6]
```diff
@@ -132,6 +137,49 @@ mod tests {
...
+                DataFrameWriteOptions::new().with_single_file_output(true),
+                Some(
+                    WriterProperties::builder()
+                        .set_compression(Compression::SNAPPY)
+                        .build(),
+                ),
+            )
+            .await?;
+
+        let read_df = ctx
```
Comment (comphead) [inline datafusion/core/src/execution/context/parquet.rs:170]: wondering what will happen on `output.parquet.snappy.parquet` ? or if there is no extension at all?

**Comment by comphead** [comment]
Thanks @Weijun-H 
Ive added some real quick minors

File: datafusion/core/src/execution/context/parquet.rs (hunk 6 above)
Comment (Weijun-H) [reply datafusion/core/src/execution/context/parquet.rs:170]: It still works as long as we provide the correct `file_extension`

**Comment by comphead** [comment]
I feel it is very close, please remove the parquet file from the PR

**Review by comphead — APPROVED** [review]
LGTM
Thanks @Weijun-H

File: datafusion/core/src/execution/context/mod.rs [hunk 7]
```diff
@@ -849,6 +849,32 @@ impl SessionContext {
...
         let session_config = self.copied_config();
         let listing_options = options.to_listing_options(&session_config);
+
+        let option_extension = listing_options.file_extension.clone();
+
+        if table_paths.is_empty() {
+            return exec_err!("No table paths were provided");
+        }
+
+        let extension = table_paths[0]
```
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:859]: I think this code is somewhat unclear if you don't have the context of this PR.

Also, why does it only look at a single file? What if the files were like

```
output1.parquet
output2.parquet.snappy
```

What would you think about extracting this logic into its own function like

```fn
     /// Heuristically determines the format (e.g. parquet, csv) to use with the `table_paths` 
     fn infer_types<'a, P: DataFilePaths>(
        table_paths: P,
    ) -> Option<String> {
...
}
``` 

A unit test could then be added that demonstrates the behavior of examples such as @comphead 's in https://github.com/apache/arrow-datafusion/pull/7972/files#r1376546426

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H

File: datafusion/core/src/execution/context/parquet.rs [hunk 8]
```diff
@@ -132,6 +137,78 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn read_from_wrong_file_extentnion() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/src/execution/context/parquet.rs:141]:
```suggestion
    async fn read_from_wrong_file_extension() -> Result<()> {
```

File: datafusion/core/src/execution/context/mod.rs [hunk 9]
```diff
@@ -849,6 +859,28 @@ impl SessionContext {
         let table_paths = table_paths.to_urls()?;
         let session_config = self.copied_config();
         let listing_options = options.to_listing_options(&session_config);
+
+        let option_extension = listing_options.file_extension.clone();
```
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:863]: I am pretty confused about this implementation. If the goal is to detect files that don't end with the expected extension, can't we just do something like

```rust
        for path in &table_paths {
            if !path.as_str().ends_with(&extension) {
                    return exec_err!(
                        "File '{path}' does not match the expected extension '{extension}'"
                    );
            }
        }
```

Comment (Weijun-H) [reply datafusion/core/src/execution/context/mod.rs:853]: Yes, you are right. I made this logic too complicated.

**Review by alamb — APPROVED** [review]
Thanks @Weijun-H  and @comphead 

Any other comments prior to merge?

**Pull request merged**